Add PSBT and wallet RPC methods

bc1cindy opened the pull request:
## Description

This PR implements the missing PSBT and wallet RPC methods that are essential for modern Bitcoin applications, particularly those using PSBT workflows like payjoin implementations.

Implemented methods:
- `wallet_create_funded_psbt` - Creates funded PSBTs with configurable options
- `wallet_process_psbt` - Signs PSBTs with wallet keys
- `finalize_psbt` - Finalizes PSBTs and extracts transactions
- `get_address_info` - Provides address validation and ownership information
- `list_unspent_with_options` - Enhanced UTXO listing with filtering capabilities

All methods follow the existing project patterns with proper error handling, type safety, and comprehensive test coverage.

### Type of Change

- [x] New feature/Enhancement (non-breaking change which adds functionality or enhances an existing one)
- [x] New or updated tests

## Notes to Reviewers

This implementation provides 100% compatibility with the methods used in rust-payjoin project, enabling them to replace their custom RPC client with `bitcoind-async-client`. All methods include comprehensive tests and follow the project's existing architectural patterns.


## Checklist

- [x] I have performed a self-review of my code.
- [x] I have commented my code where necessary.
- [x] I have updated the documentation if needed.


## Related Issues

- Closes #31
- Related to payjoin/rust-payjoin#945

storopoli commented on src/types.rs:
```diff
@@ -883,3 +883,57 @@ where
+pub struct WalletCreateFundedPsbtOptions {
+    #[serde(rename = "fee_rate", skip_serializing_if = "Option::is_none")]
+    pub fee_rate: Option<f64>,
+    #[serde(rename = "lock_unspent", skip_serializing_if = "Option::is_none")]
+    pub lock_unspent: Option<bool>,
```
this one is camelCase at least in bitcoin core 29.0: https://bitcoincore.org/en/doc/29.0.0/rpc/wallet/walletcreatefundedpsbt/

```suggestion
    #[serde(rename = "lockUnspents", skip_serializing_if = "Option::is_none")]
    pub lock_unspents: Option<bool>,
```

Also note that we have plural, so it's `lockUnspents` (check the `s`)

storopoli commented on src/types.rs:
```diff
@@ -883,3 +883,57 @@ where
+    pub replaceable: Option<bool>,
+}
+
+#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
+pub struct WalletCreateFundedPsbt {
+    pub psbt: String,
+    #[serde(deserialize_with = "deserialize_bitcoin")]
+    pub fee: Amount,
+    #[serde(rename = "changepos")]
+    pub change_pos: i32,
+}
```
Can we add `locktime` as well?
We don't have a `deserialize_locktime` but you are free to include, otherwise we can tackle in a follow-up PR. If you don't want to add it please open a ticket for this?
If you want to go ahead just note that this might be tricky because you'll be deserializing an unsigned int as either an `{relative,absolute}::LockTime` (see https://docs.rs/bitcoin/latest/bitcoin/absolute/enum.LockTime.html and https://docs.rs/bitcoin/latest/bitcoin/relative/enum.LockTime.html) and you'll need to have the proper safeguards (and tests!)

storopoli commented on src/types.rs:
```diff
@@ -883,3 +883,57 @@ where
+    pub change_pos: i32,
+}
+
+#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
+pub struct WalletProcessPsbt {
+    pub psbt: String,
+    pub complete: bool,
+}
```
We are missing `sighashtype`. I don't care much to `sign` and `bip32derivs`.
If you don't want to implement `sighashtype` please open an issue?

storopoli commented on src/types.rs:
```diff
@@ -883,3 +883,57 @@ where
+    pub complete: bool,
+}
+
+#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
+pub struct FinalizePsbt {
+    pub psbt: Option<String>,
+    pub hex: Option<String>,
+    pub complete: bool,
+}
```
NACK, this should be a `WalletProcessPsbtResult` similar to https://github.com/alpenlabs/bitcoind-async-client/blob/c151abd31f8d47d2ba76282da56217d5fe725827/src/types.rs#L359-L376

storopoli commented on src/types.rs:
```diff
@@ -883,3 +883,57 @@ where
+    pub solvable: Option<bool>,
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize)]
+pub struct ListUnspentQueryOptions {
+    #[serde(rename = "minimumAmount", skip_serializing_if = "Option::is_none")]
+    pub minimum_amount: Option<f64>,
+    #[serde(rename = "maximumAmount", skip_serializing_if = "Option::is_none")]
+    pub maximum_amount: Option<f64>,
+    #[serde(rename = "maximumCount", skip_serializing_if = "Option::is_none")]
+    pub maximum_count: Option<u32>,
+}
```
NACK: all these `rename_with` can be replaced by a single `#[serde(rename_all = "camelCase")]` in the `struct`.
See: https://serde.rs/attr-rename.html

```suggestion
#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ListUnspentQueryOptions {
    #[skip_serializing_if = "Option::is_none")]
    pub minimum_amount: Option<f64>,
    #[skip_serializing_if = "Option::is_none")]
    pub maximum_amount: Option<f64>,
    #[skip_serializing_if = "Option::is_none")]
    pub maximum_count: Option<u32>,
}
```

storopoli commented on src/types.rs:
```diff
@@ -883,3 +883,57 @@ where
+
+#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
+pub struct WalletCreateFundedPsbt {
+    pub psbt: String,
```
No strings! See: https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/
We should have a function `deserialize_psbt` similar to https://github.com/alpenlabs/bitcoind-async-client/blob/c151abd31f8d47d2ba76282da56217d5fe725827/src/types.rs#L777-L801

storopoli commented on src/types.rs:
```diff
@@ -883,3 +883,57 @@ where
+
+#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
+pub struct GetAddressInfo {
+    pub address: String,
```
No strings! See: https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/
Use https://github.com/alpenlabs/bitcoind-async-client/blob/c151abd31f8d47d2ba76282da56217d5fe725827/src/types.rs#L803-L833

storopoli commented on src/traits.rs:
```diff
@@ -245,4 +270,18 @@ pub trait Signer {
+        &self,
+        psbt: &str,
+        sign: Option<bool>,
+        sighashtype: Option<String>,
```
Nit! No strings! We can use [`TapSighashType`](https://docs.rs/bitcoin/latest/bitcoin/sighash/enum.TapSighashType.html) and deal with the `0` that is not covered by [`EcdsaSighashType`](https://docs.rs/bitcoin/latest/bitcoin/sighash/enum.EcdsaSighashType.html) or create new `enum` in `types.rs` to deal with all sighashes from ECDSA and Taproot.

storopoli reviewed (changes requested):
Hey, amazing, thanks for the PR! I have a couple of suggestions. Please let me know if you intend to address them or if someone else should take over this PR.

A lot of the PSBT tests can be included in the `client_works` so that we don't have to mine a lot of 101 blocks in a virgin retest everytime:
https://github.com/alpenlabs/bitcoind-async-client/blob/c151abd31f8d47d2ba76282da56217d5fe725827/src/client.rs#L580
Can you try putting them in there? And which ones you cannot we can leave as standalone tests.

bc1cindy commented:
> Hey, amazing, thanks for the PR! I have a couple of suggestions. Please let me know if you intend to address them or if someone else should take over this PR.
> 
> A lot of the PSBT tests can be included in the `client_works` so that we don't have to mine a lot of 101 blocks in a virgin regtest everytime:
> 
> https://github.com/alpenlabs/bitcoind-async-client/blob/c151abd31f8d47d2ba76282da56217d5fe725827/src/client.rs#L580
> 
> 
> Can you try putting them in there? And which ones you cannot we can leave as standalone tests.
> EDIT: `clippy` is failing on CI, ignore the `zizmor` failure since I'm tracking in #33.

Hi @storopoli, thank you for the thorough review! I really appreciate the detailed feedback.

I understand all the points and will address them all, including locktime and sighashtype implementation.

storopoli commented on src/types.rs:
```diff
@@ -1,11 +1,11 @@
 
 use bitcoin::{
-    absolute::Height,
+    absolute::{Height, LockTime},
```
EDIT: We don't need this diff anymore

storopoli commented on src/client.rs:
```diff
@@ -1,4 +1,5 @@
 use std::{
+    collections::HashMap,
```
With `CreateRawTransactionOutput` we don't need `HashMap`s

storopoli commented on src/client.rs:
```diff
@@ -731,6 +819,99 @@ mod test {
         assert_eq!(expected, got);
+
+        let psbt_address = client.get_new_address().await.unwrap();
+        let mut psbt_outputs = std::collections::HashMap::new();
+        psbt_outputs.insert(psbt_address.to_string(), 1.0);
```
With `CreateRawTransactionOutput` we don't need `HashMap`s

storopoli commented on src/client.rs:
```diff
@@ -731,6 +819,99 @@ mod test {
+        mine_blocks(&bitcoind, 1, None).unwrap();
+
+        let utxos = client
+            .list_unspent_with_options(Some(1), Some(9999999), None, Some(true), None)
```
Please add `_` thousands separator
```suggestion
            .list_unspent_with_options(Some(1), Some(9_999_999), None, Some(true), None)
```

storopoli commented on src/client.rs:
```diff
@@ -731,6 +819,99 @@ mod test {
+        let utxos_filtered = client
+            .list_unspent_with_options(
+                Some(1),
+                Some(9999999),
```
Please add `_` thousands separator

```suggestion
                Some(9_999_999),
```

storopoli commented on src/client.rs:
```diff
@@ -731,6 +819,99 @@ mod test {
+        let utxos_with_query = client
+            .list_unspent_with_options(
+                Some(1),
+                Some(9999999),
```
Please add `_` thousands separator

```suggestion
                Some(9_999_999),
```

storopoli commented on src/traits.rs:
```diff
@@ -209,6 +211,29 @@ pub trait Wallet {
         raw_tx: CreateRawTransaction,
     ) -> impl Future<Output = ClientResult<Transaction>> + Send;
+
+    fn wallet_create_funded_psbt(
+        &self,
+        inputs: &[CreateRawTransactionInput],
+        outputs: &HashMap<String, f64>,
+        locktime: Option<u32>,
+        options: Option<WalletCreateFundedPsbtOptions>,
+        bip32_derivs: Option<bool>,
+    ) -> impl Future<Output = ClientResult<WalletCreateFundedPsbt>> + Send;
```
docstring the functions following the convention in this file/crate.

storopoli commented on src/traits.rs:
```diff
@@ -209,6 +211,29 @@ pub trait Wallet {
+        bip32_derivs: Option<bool>,
+    ) -> impl Future<Output = ClientResult<WalletCreateFundedPsbt>> + Send;
+
+    fn get_address_info(
+        &self,
+        address: &Address,
+    ) -> impl Future<Output = ClientResult<GetAddressInfo>> + Send;
```
docstring the functions following the convention in this file/crate.

storopoli commented on src/traits.rs:
```diff
@@ -209,6 +211,29 @@ pub trait Wallet {
+        address: &Address,
+    ) -> impl Future<Output = ClientResult<GetAddressInfo>> + Send;
+
+    fn list_unspent_with_options(
+        &self,
+        min_conf: Option<u32>,
+        max_conf: Option<u32>,
+        addresses: Option<&[Address]>,
+        include_unsafe: Option<bool>,
+        query_options: Option<ListUnspentQueryOptions>,
+    ) -> impl Future<Output = ClientResult<Vec<ListUnspent>>> + Send;
```
docstring the functions following the convention in this file/crate.

storopoli reviewed (changes requested):
This looks almost good to go. But I fear we need another round.
Also can you please add docstrings to everything following the convention/idiosyncrasy of the codebase?

Finally, aren't you guys gonna need [`psbtbumpfee`](https://bitcoincore.org/en/doc/29.0.0/rpc/wallet/psbtbumpfee/) in Payjoin? This is not included in this PR or the related issue Cc @DanGould.

EDIT: To avoid failing CI you can run the CI locally with `cargo clippy` and `cargo fmt` etc. before pushing.

bc1cindy commented:
@storopoli I've addressed all the feedback from your review:

- Added docstrings everywhere following the project conventions
- Switched from HashMap to CreateRawTransactionOutput for type safety
- Fixed WalletProcessPsbtResult.hex to deserialize as Transaction instead of String
- Removed the locktime field from WalletCreateFundedPsbt
- Updated the test to use the proper BIP 174 test vector you suggested
- Fixed the SighashType import to use the full path
- Added thousands separators where you pointed them out
- Implemented psbtbumpfee

Everything is passing locally, all tests with both cargo test and nextest, plus clippy and formatting checks are clean.

Ready for re-review.

storopoli reviewed (changes requested):
Really good almost there! A few requested changes, let me know if you have any issues with these suggestions.

EDIT: `cargo clippy` is still failing please make sure to run it before committing new changes.

DanGould commented:
> aren't you guys gonna need psbtbumpfee in Payjoin? This is not included in this PR or the related issue

We don't use psbtbumpfee in the reference implementation. I suppose we could in the future, but it may be ok to leave for a follow up PR.

bc1cindy commented:
Thanks for the excellent and thorough code review @storopoli 

I've implemented all requested changes:

-  Renamed list_unspent_with_options → list_unspent
-  Added empty lines between all struct fields/enum variants
-  Removed duplicate PsbtBumpFeeOutput → using existing CreateRawTransactionOutput
-  Converted test functions → constants (const TEST_PSBT, const TEST_TX_HEX)
-  Added missing docstrings to WalletProcessPsbtResult
-  Implemented standalone psbt_bump_fee test
-  Fixed clippy::cloned-ref-to-slice-refs error → std::slice::from_ref
-  Updated Cargo.lock with "base64" feature

 Validation:

- cargo clippy passing
- All tests working
- Manually tested with Bitcoin Core v29.0

storopoli reviewed (approved):
ACK 9734de580d03bd6e45e5028c8d881417f24e447b

storopoli reviewed (approved):
ACK 8ad5ffa8d52405b5cfc5df7e96e49064b71c9459

bc1cindy commented:
> @bc1cindy I did some changes in [9734de5](https://github.com/alpenlabs/bitcoind-async-client/commit/9734de580d03bd6e45e5028c8d881417f24e447b) and in [8ad5ffa](https://github.com/alpenlabs/bitcoind-async-client/commit/8ad5ffa8d52405b5cfc5df7e96e49064b71c9459) will merge as it is. Thank you for the amazing PR.

YAY, thank you!! @storopoli! 🙏

Really appreciate your excellent review and expertise in getting this merged. Learned a lot from your changes, great collaboration on my first contribution to this project!
